### ISSUE TITLE
Unable to add CSS class names to editor iframe

### ISSUE THREAD

**Issue by markhowellsmead** [WordPress/gutenberg#56831]
### Description

I'm attempting to add CSS class names to the new iframe in the editor using JavaScript. The aim is to apply a user-selected accent colour in the editor.

I've only been able to do so using an overly-complicated combination of MutationObserver (to detect when the iframe becomes available) and a setInterval to repeatedly try to apply the class name until it sticks. This seems to be unnecessarily complex. Is there a better solution?

I've tried to add the class names using the `admin_body_class` hook in PHP, but this doesn't feel right (as it's PHP) and it also only appears to work on the main `body` (which doesn't help, as it's hidden by the iframe).

### Step-by-step reproduction instructions

See the code example.

### Screenshots, screen recording, code snippet

https://gist.github.com/markhowellsmead/01c8c29937162e93a11e21c8209872d1

### Environment info

- WordPress 6.4
- No plugins.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by markhowellsmead** [comment]
See #17854 and https://github.com/WordPress/gutenberg/blob/cad8faebd613d5e821fe7f1eaf568fa4b56b6f88/packages/block-editor/src/components/iframe/index.js#L144-L152.

The solution is to use the `admin_body_class` hook in PHP to add class names to the main `body`. By prefixing the class names with `admin-color-`, the class names will be automatically copied over to the editor iframe.

**Comment by markhowellsmead** [comment]
Reopening as there should be a more logical way to do this.

**Comment by ramonjd** [comment]
I'm just throwing this out there without much thought, so it might be irrelevant or useless.

The comment above says:

>  // Ideally ALL classes that are added through get_body_class should 
>  // be added in the editor too, which we'll somehow have to get from 
>  // the server in the future (which will run the PHP filters).

I wonder if it'd be okay to inject the return value of [body_class()](https://github.com/WordPress/wordpress-develop/blob/52854e2796bff1ab2e69991b1c233c277542a0fc/src/wp-includes/post-template.php#L618) in `get_block_editor_settings`, whose output is rendered in the editor HTML — and therefore accessible to the block editor applications —  via `wp.[editSite | editPost].initializeEditor`?

**Comment by colinduwe** [comment]
Here's a proposal PR. I'd appreciate any feedback https://github.com/WordPress/gutenberg/pull/68538

**Comment by colinduwe** [comment]
@Mamaduka has [suggested](https://github.com/WordPress/gutenberg/issues/68541) that we should not use $settings and `get_block_editor_settings` to add a bodyClass property to the editor. I am led to believe that is a deprecated method for transmitting data to the editor. What is the current best practice for sending a new property like `bodyClasses` to the editor from the server that also supports a PHP filter so extenders can then modify that property before the editor initializes?

Once the property is part of the 'core/block-editor' state, it becomes pretty easy to hook in and keep the body classes up to date.

**Comment by Mamaduka** [comment]
Thanks, @colinduwe!

I must look up why iframe limits classes copied from the parent body. The original code was introduced in #32057, but I don't see any related comments. My current guess is to avoid leaking styles into the iframe.

Technically speaking, I don't see a difference between injecting values returned by `body_class()` via selector or letting iframe copy all the classes.

Code in question:

https://github.com/WordPress/gutenberg/blob/cad8faebd613d5e821fe7f1eaf568fa4b56b6f88/packages/block-editor/src/components/iframe/index.js#L144-L152

**Comment by colinduwe** [comment]
@Mamaduka The current iframe body class implementation isn't flexible enough. It currently only runs onLoad and the classes are proscribed there in the onLoad. Instead, the bodyClasses should be a part of the blockEditorStore. Have a look [here](https://github.com/WordPress/gutenberg/pull/68538/files#diff-ab17abd38e687edd56573694ae3017e105e96b319571c190e9d4705c1872bc9cL142-R160).
That way we can update the bodyClasses, for example [when the user swaps templates](https://github.com/WordPress/gutenberg/pull/68538/files#diff-22558edfc3d13321867c7bfcc5ba7a223097f027becd20a71d0fb409c39d7eb7).

Currently, body classes that we had in the TinyMCE era are not present in the ownerDocument.body.classlist. For example, page-template.

I'd also like to call attention to https://github.com/WordPress/gutenberg/issues/17854 begun in 2019. Is there a single issue where we should be discussing this?

**Comment by markhowellsmead** [comment]
Agreed that this should be subscribable. The point of the origins requirement is that the frame colour can be changed at any time when clicking on a page-level colour selector.

**Comment by Mamaduka** [comment]
> Is there a single issue where we should be discussing this?

Yes, it would be nice to consolidate this into a single issue. The usual policy is to use the oldest issue, reference new ones and close them.

Somewhat related: https://github.com/WordPress/gutenberg/issues/60299#issuecomment-2585721484.

**Comment by afercia** [comment]
An objection to close this issue in favor of https://github.com/WordPress/gutenberg/issues/17854 ?

**Comment by afercia** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/issues/17854. Can always be reopened if needed.
